OsmAnd/src/net/osmand/plus/views/layers/POIMapLayer.java

```diff
@@ -173,24 +175,17 @@ public boolean isCancelled() {
 									return isInterrupted();
 								}
 							});
-					if (filter.isRouteTrackFilter()) {
-						for (Amenity amenity : amenities) {
-							boolean hasRoute = false;
+					for (Amenity amenity : amenities) {
+						if (amenity.isRouteTrack()) {
 							String routeId = amenity.getRouteId();
-							if (!Algorithms.isEmpty(routeId)) {
-								for (Amenity a : res) {
-									if (routeId.equals(a.getRouteId())) {
-										hasRoute = true;
-										break;
-									}
+							if (routeId != null) {
+								if (deduplicateByRouteId.contains(routeId)) {
```

Comment:
Probably we can try to make a more compact check, something like this - `for (Amenity amenity : amenities) {
						if (amenity.isRouteTrack()) {
							String routeId = amenity.getRouteId();
							if (routeId != null && !uniqueRouteIds.add(amenity.getRouteId())) {
								continue;
							}
						}
						res.add(amenity);
					}`

Reply:
62134bf41d